prov/rxd/src/rxd_ep.c

```diff
@@ -743,6 +781,108 @@ static int rxd_dg_cq_open(struct rxd_ep *rxd_ep, enum fi_wait_obj wait_obj)
 	return ret;
 }
 
+int rxd_epaddr_hash_insert(struct rxd_ep *ep, void *addr, 
+				fi_addr_t *rxd_addr)
+{
+	struct rxd_epname_entry *entry = NULL;
+
+	HASH_FIND(hh, ep->ep_rxdaddr_hash, addr, RXD_NAME_LENGTH, entry);
+	if (entry != NULL) {
+		return -FI_EALREADY;
+	}
+	entry = calloc(1, sizeof(*entry));
+	assert(entry && rxd_addr);
+	memcpy(entry->addr, addr, RXD_NAME_LENGTH);
+	entry->rxd_addr = *rxd_addr;
+	HASH_ADD(hh, ep->ep_rxdaddr_hash, addr, RXD_NAME_LENGTH, entry);
+	return 0;
+}
+
+struct rxd_peer* rxd_get_peer_entry(struct rxd_ep *ep, const void *addr, size_t addrlen) 
+{
+	struct rxd_epname_entry *entry = NULL;
+	struct rxd_peer *peer_entry;
+	fi_addr_t rxd_addr; 
+	uint8_t tmp_addr[RXD_NAME_LENGTH];
+
+	memset(tmp_addr, 0, RXD_NAME_LENGTH);
+	memcpy(tmp_addr, addr, addrlen);
+	HASH_FIND(hh, ep->ep_rxdaddr_hash, tmp_addr, RXD_NAME_LENGTH, entry);
+	if (entry)
+		return ofi_bufpool_get_ibuf(ep->peer_pool.pool, entry->rxd_addr);
+	peer_entry = ofi_ibuf_alloc(ep->peer_pool.pool);
+	rxd_addr = ofi_buf_index(peer_entry); 
+	rxd_epaddr_hash_insert(ep, tmp_addr, &rxd_addr);
+	return peer_entry;
+}
+
+int rxd_fiaddr_hash_insert(struct rxd_ep *ep, fi_addr_t *fi_addr, 
+				fi_addr_t *rxd_addr)
+{
+	struct rxd_fiaddr_entry *entry = NULL;
+	HASH_FIND(hh, ep->fi_rxdaddr_hash, (void*)fi_addr, 
+			sizeof(*fi_addr), entry);
+	if (entry != NULL)
+		return -FI_EALREADY;
+	entry = malloc(sizeof(*entry));
+	assert(entry && fi_addr && rxd_addr);
+	entry->fi_addr = *fi_addr;
+	entry->rxd_addr = *rxd_addr;
+	HASH_ADD(hh, ep->fi_rxdaddr_hash, fi_addr, sizeof(*fi_addr), entry);
+	return 0;
+}
+void rxd_fi_update_peer_entry(struct rxd_ep *ep, const void* addr,
+				fi_addr_t fi_addr, fi_addr_t *dg_addr, size_t addrlen)
+{
+	struct rxd_peer *peer_entry;
+	fi_addr_t rxd_addr;
+	
+	peer_entry = rxd_get_peer_entry(ep, addr, addrlen);
+	peer_entry->fi_addr = fi_addr;  
+	peer_entry->dg_addr = *dg_addr;
+	rxd_addr = ofi_buf_index(peer_entry);
+
+	if (rxd_fiaddr_hash_insert(ep, &fi_addr, &rxd_addr))
+		FI_WARN(&rxd_prov, FI_LOG_AV, "Unable to create fi_addr mapping\n");
+}
+
+int rxd_update_av_peers(struct util_av *util_av, void *dg_addr, 
+				fi_addr_t fi_addr, void *arg)
+{	
+	struct rxd_av *av;
+	struct rxd_ep *ep;
+	int ret;
+	uint8_t addr[RXD_NAME_LENGTH];
+	size_t addrlen;
+	addrlen = RXD_NAME_LENGTH;
+	ep = (struct rxd_ep*)arg;
+	av = rxd_ep_av(ep);
+
+	memset(addr, 0, RXD_NAME_LENGTH);
+	fastlock_acquire(&av->util_av.lock);
+	ret = fi_av_lookup(av->dg_av, *((fi_addr_t*)dg_addr), addr, &addrlen);
+	if (ret) {
+		FI_WARN(&rxd_prov, FI_LOG_AV, "Unable to find inserted dg address\n");
+		goto err;
+	}
+	rxd_fi_update_peer_entry(ep, addr, fi_addr, (fi_addr_t*)dg_addr,
```

Comment:
The use of the word "update" here is a bit confusing and misleading. The peer can be initialized but the word update suggests that you are changing information from a peer that was already initialized. Consider renaming and reorganizing the functions rxd_fi_update_peer_entry, rxd_update_av_peers, and rxd_get_peer_entry to clarify which ones are doing initialization, updating, or getting. Otherwise, it's unclear where the peer is actually getting created/updated/looked up.